File: src/serverless/BlazingPizza.Functions.FSharp/PlaceOrder.fs
namespace BlazingPizza.Functions.FSharp

open Microsoft.AspNetCore.Mvc
open Microsoft.Azure.WebJobs
open Microsoft.Azure.WebJobs.Extensions.Http
open Microsoft.AspNetCore.Http
open Microsoft.Extensions.Logging

open BlazingPizza
open System.IO
open Newtonsoft.Json
open System

module PlaceOrder =
    let validatePizza (sp: Pizza) (pizzas: PizzaSpecial seq) =
        let pizza =
            pizzas
            |> Seq.tryFind (fun p -> p.Id = sp.Special.Id)

        match pizza with
        | Some pizza ->
            sp.SpecialId <- pizza.Id
            sp.Special <- pizza
        | None -> failwith "You tried to submit a pizza we don't make"

    let validateToppings (toppings: Topping seq) (st: PizzaTopping) =
        let topping =
            toppings
            |> Seq.tryFind (fun t -> t.Id = st.Topping.Id)

        match topping with
        | Some topping ->
            st.ToppingId <- topping.Id
            st.Topping <- topping
        | None -> failwith "You tried to submit a topping we don't have"

    [<FunctionName("place-orders")>]
    let run
        ([<HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = null)>] req: HttpRequest)
        ([<CosmosDB("blazingPizza",
                    "pizza",
                    ConnectionStringSetting = "CosmosConnectionString",
                    SqlQuery = "SELECT * FROM c WHERE c.partitionKey = 'specials'")>] pizzas: PizzaSpecial seq)
        ([<CosmosDB("blazingPizza",
                    "pizza",
                    ConnectionStringSetting = "CosmosConnectionString",
                    SqlQuery = "SELECT * FROM c WHERE c.partitionKey = 'topping'")>] toppings: Topping seq)
        ([<CosmosDB("blazingPizza", "pizza", ConnectionStringSetting = "CosmosConnectionString")>] orders: IAsyncCollector<Order>)
        (log: ILogger)
        =
        async {
            use sr = new StreamReader(req.Body)
            let! rawBody = sr.ReadToEndAsync() |> Async.AwaitTask

            let order =
                JsonConvert.DeserializeObject<Order> rawBody

            order.CreatedTime <- DateTime.Now
            order.DeliveryLocation <- LatLong(51.5001, -0.1239)
            order.UserId <- "Mr Awesome"

            let validateToppings' = validateToppings toppings
            order.Pizzas
            |> Seq.iter (fun sp ->
                validatePizza sp pizzas
                sp.Toppings |> Seq.iter validateToppings')

            let random = Random()
            order.OrderId <- random.Next()

            do! orders.AddAsync(order) |> Async.AwaitTask

            return OkObjectResult(order.OrderId)
        }
        |> Async.StartAsTask

File: src/serverless/BlazingPizza.Functions.CSharp/PlaceOrder.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Logging;
using System.Collections.Generic;
using System.Linq;
using System.IO;
using System.Threading.Tasks;
using Newtonsoft.Json;
using System;
using System.Security.Claims;

namespace BlazingPizza.Functions.CSharp
{
    public static class PlaceOrder
    {
        private static void ValidatePizza(Pizza sp, IEnumerable<PizzaSpecial> pizzas)
        {
            var pizza = pizzas.FirstOrDefault(p => p.Id == sp.Special.Id);

            if (pizza == default(PizzaSpecial))
            {
                throw new KeyNotFoundException("You tried to submit a pizza we don't make");
            }

            sp.SpecialId = pizza.Id;
            sp.Special = pizza;
        }

        private static void ValidateToppings(IEnumerable<Topping> toppings, PizzaTopping st)
        {
            var topping = toppings.FirstOrDefault(t => t.Id == st.Topping.Id);

            if (topping == default(Topping))
            {
                throw new KeyNotFoundException("You tried to submit a topping we don't have");
            }

            st.ToppingId = topping.Id;
            st.Topping = topping;
        }

        [FunctionName("orders")]
        public static async Task<IActionResult> Run(
            [HttpTrigger(AuthorizationLevel.Function, "post", Route = null)] HttpRequest req,
            [CosmosDB("blazingPizza",
                      "pizza",
                      ConnectionStringSetting = "CosmosConnectionString",
                      SqlQuery = "SELECT * FROM c WHERE c.partitionKey = 'specials'")] IEnumerable<PizzaSpecial> pizzas,
            [CosmosDB("blazingPizza",
                      "pizza",
                      ConnectionStringSetting = "CosmosConnectionString",
                      SqlQuery = "SELECT * FROM c WHERE c.partitionKey = 'topping'")] IEnumerable<Topping> toppings,
            [CosmosDB("blazingPizza", "pizza", ConnectionStringSetting = "CosmosConnectionString")] IAsyncCollector<Order> orders,
            ILogger log)
        {
            using (var sr = new StreamReader(req.Body))
            {
                var rawBody = await sr.ReadToEndAsync();
                var order = JsonConvert.DeserializeObject<Order>(rawBody);

                order.CreatedTime = DateTime.Now;
                order.DeliveryLocation = new LatLong(51.5001, -0.1239);

                var principal = StaticWebAppsAuth.Parse(req);
                var userId = principal.Claims.First(c => c.Type == ClaimTypes.NameIdentifier);
                order.UserId = userId.Value;

                foreach (var pizza in order.Pizzas)
                {
                    ValidatePizza(pizza, pizzas);

                    foreach (var topping in pizza.Toppings)
                    {
                        ValidateToppings(toppings, topping);
                    }
                }

                order.OrderId = new Random().Next();

                await orders.AddAsync(order);

                return new OkObjectResult(order.OrderId);
            }
        }
    }
}
